server/storage/mvcc/watchable_store.go

```diff
@@ -370,6 +370,12 @@ func (s *watchableStore) syncWatchers() int {
 	victims := make(watcherBatch)
 	wb := newWatcherBatch(wg, evs)
 	for w := range wg.watchers {
+		if w.minRev < compactionRev {
+			// skip the watcher that failed to send compacted watch response due to w.ch is full
+			// next retry of syncWatchers would try to resend the compacted watch response to w.ch
+			// TODO prioritize sending compacted watch response over other watch responses with events.
+			continue
+		}
```

Comment:
I think we make the fix better if we move it up higher. Let's group stages of this code:
* Line 354: Pick unsynched watchers that we want to sync, and their minRev which should be higher than compactionRev. It might also drop compacted  watchers.
* Lines 355-368:  Read events between minRev and curRev
* Line 371: Match watchers to events they should receive
* Lines 372+: Pass events to watchers

By implementing fix on line 373, you are skipping sync of watch that we already matched events too. 

Instead I recommend to change the Line 354 to not return compacted watches, even if `w.ch` is full. Don't know if you tried it already, but on first try it will not work properly, why? Because the `.choose` function is written badly. It sometimes mutates the original watchGroup, sometimes returns the same watch group, sometimes a copy. It's not great.

My recommended implementation of choose and chooseAll (also recommend rename of chooseAll to `minRev`) 
```
// choose selects watchers from the watcher group to update
func (wg *watcherGroup) choose(maxWatchers int, curRev, compactRev int64) (*watcherGroup, int64) {
	watcherNeedsCompaction := false
	for w := range wg.watchers {
		if w.minRev < compactRev {
			watcherNeedsCompaction = true
		}
	}
	if len(wg.watchers) < maxWatchers && !watcherNeedsCompaction {
		return wg, wg.chooseAll(curRev, compactRev)
	}
	ret := newWatcherGroup()
	for w := range wg.watchers {
		if maxWatchers <= 0 {
			break
		}
		maxWatchers--
		if w.minRev < compactRev {
			select {
			case w.ch <- WatchResponse{WatchID: w.id, CompactRevision: compactRev}:
				w.compacted = true
				wg.delete(w)
			default:
				// retry next time
			}
			continue
		}
		ret.add(w)
	}
	return &ret, ret.chooseAll(curRev, compactRev)
}

func (wg *watcherGroup) chooseAll(curRev, compactRev int64) int64 {
	minRev := int64(math.MaxInt64)
	for w := range wg.watchers {
		if w.minRev > curRev {
			// after network partition, possibly choosing future revision watcher from restore operation
			// with watch key "proxy-namespace__lostleader" and revision "math.MaxInt64 - 2"
			// do not panic when such watcher had been moved from "synced" watcher during restore operation
			if !w.restore {
				panic(fmt.Errorf("watcher minimum revision %d should not exceed current revision %d", w.minRev, curRev))
			}

			// mark 'restore' done, since it's chosen
			w.restore = false
		}
		if minRev > w.minRev && minRev > compactRev {
			minRev = w.minRev
		}
	}
	return minRev
}
```

I think it's cleaner, as for performance I think we need to test as there is a tradeoff between copying `watchableStore` or matching events.

Reply:
Thanks for the feedback! 

> Line 354: Pick unsynched watchers that we want to sync, and their minRev which should be higher than compactionRev. It might also drop compacted watchers.

I agree it's a cleaner fix because we should not update the compacted watchers in the first place in Line 354. Currently, line 371 `wb := newWatcherBatch(wg, evs)` is a waste of space allocated up to N (compacted watchers) * `watchBatchMaxRevs` (1000) events and then garbage collected. 

> as for performance I think we need to test as there is a tradeoff between copying watchableStore or matching events. 

Don't understand this statement because I think we are already copying `watcherGroup` if # of unsync watchers is greater than 512. 

My question is: 

Is this recommended change acceptable for backporting 3.5 & 3.4 release branch comparing to original 2 lines of code change? Both proposed changes look correct to me.  @ahrtr @serathius @fuweid

Reply:
I think this is a better fix as it is more readable, we can clearly separate responsibility of:
* Sending compacted response to watcher and removing it
* Selecting list of watchers we want to synchronize
* Picking the minimal revision between members we want to synchronize. 

Don't have anything against backporting this to older releases. I think it's better to backport a proper fix, than backporting just a shortcut, leaving a proper fix just on main branch, causing branches to diverge and further backports harder.

Reply:
just my two cents. even if we don't return watchers that they need to be compacted, we still need condition checker to make sure please do not touch their watchers' minRev. For the waiterGroup bug, we can have separate commits (in one pull request) to fix. The v3.4 release is still active and codebase layout is different from main and 3.5. multiple commits could help when we backport. Just don't want to involve refactor things in bugfix backports.

Reply:
> Is this recommended change acceptable for backporting 3.5 & 3.4 release branch comparing to original 2 lines of code change? Both proposed changes look correct to me. @ahrtr @serathius @fuweid

Usually we do code refactoring on main branch only. But I am not against doing such kind of local/small area of code refactoring even in release-3.4/5, because we have to fix the second related issue as mentioned in https://github.com/etcd-io/etcd/pull/17555#issuecomment-1985854328.

Please deliver a PR and we will evaluate the impact. Thanks.

Reply:
Thanks all, let's keep the current fix as it is and would refactor and fix the https://github.com/etcd-io/etcd/pull/17555#issuecomment-1985854328 in follow up PRs.